### ISSUE TITLE
pointer being freed was not allocated

### ISSUE THREAD

**Issue by rened** [JuliaLang/julia#10829]
Starting with commit 11a0f43cf322e24c96c7f64b82534187c962c052 (from git bisect) a simple `Pkg.add("HDF5")` fails on a fresh Julia build with no packages installed, on both OSX and Ubuntu.

#### OSX

```
100  258k  100  258k    0     0  80968      0  0:00:03  0:00:03 --:--:-- 80987
==> Tapping staticfloat/juliadeps
Tapped 31 formulae (55 files, 452K)
HEAD is now at 28d83a8 gobject-introspection: update 1.44.0 bottle.
HEAD is now at eb0a89f Merge pull request #42 from staticfloat/staging
INFO: Building HDF5
julia(98697,0x7fff78431310) malloc: *** error for object 0x300007fc81d65b23: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
```
```
Julia Version 0.4.0-dev+4214
Commit 11a0f43* (2015-04-14 03:20 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```
#### Ubuntu

```
*** Error in `./julia': double free or corruption (fasttop): 0x0000000003503c80 ***

signal (6): Aborted
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: -1375857772)
unknown function (ip: -1375807890)
jl_uv_dlopen at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
check_path! at /home/rene/.julia/v0.4/BinDeps/src/dependencies.jl:530
jlcall_check_path!_44548 at  (unknown line)
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
_find_library at /home/rene/.julia/v0.4/BinDeps/src/dependencies.jl:515
satisfy! at /home/rene/.julia/v0.4/BinDeps/src/dependencies.jl:735
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
satisfy! at /home/rene/.julia/v0.4/BinDeps/src/dependencies.jl:735
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
anonymous at /home/rene/.julia/v0.4/BinDeps/src/dependencies.jl:790
unknown function (ip: -1368847253)
unknown function (ip: -1368844575)
jl_load at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
include at ./boot.jl:250
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
include_from_node1 at ./loading.jl:129
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
unknown function (ip: -1368924453)
unknown function (ip: -1368927294)
unknown function (ip: -1368847469)
unknown function (ip: -1368847315)
jl_toplevel_eval_in at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
evalfile at loading.jl:171
evalfile at loading.jl:171
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
anonymous at pkg/entry.jl:626
cd at ./file.jl:20
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
build! at pkg/entry.jl:625
jlcall_build!_44303 at  (unknown line)
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
build at pkg/entry.jl:637
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
resolve at ./pkg/entry.jl:446
edit at pkg/entry.jl:24
jl_apply_generic at /home/rene/local/devdevjulia/usr/bin/../lib/libjulia.so (unknown line)
anonymous at task.jl:361
unknown function (ip: -1368897952)
unknown function (ip: 0)
Aborted (core dumped)
```
```
Julia Version 0.4.0-dev+4214
Commit 11a0f43* (2015-04-14 03:20 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7 CPU         920  @ 2.67GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

cc @garrison

**Comment by MillCheck** [comment]
I have had problems with HDF5 in the past, like the one where Blosc didn't build, starting julia as superuser helped for some reason, but this seems like a diffrent problem altogether :/

**Comment by rened** [comment]
Sorry, I should have added that this is not `HDF5` specific - it happens with `Images` as well. I am quite positive that it has nothing to do with the package itself.

**Comment by ihnorton** [comment]
See https://github.com/JuliaLang/BinDeps.jl/pull/140 -- you could test with `Pkg.checkout("BinDeps")`

**Comment by rened** [comment]
Unfortunately still failing with BinDeps@aac14b45949a95b812114482808d084251b85408 on both OSX and Linux

**Comment by ihnorton** [comment]
Yes, looks like it is different. I'm suspicious of the interaction of [this bit](https://github.com/JuliaLang/BinDeps.jl/blob/aac14b45949a95b812114482808d084251b85408/src/dependencies.jl#L471-L477) but can't see an obviously bad path yet.. Should be pretty easy to find with rr if someone has that set up.

**Comment by garrison** [comment]
[This line](https://github.com/JuliaLang/BinDeps.jl/blob/aac14b45949a95b812114482808d084251b85408/src/dependencies.jl#L529) should be a `calloc` instead of malloc, since the new commit assumes `handle->errmsg` is initialized to NUL.  Although there might be an even better way to fix this.

**Comment by garrison** [comment]
In fact, I think the better idea is to have `jl_uv_dlopen` make no assumptions about whether anything is null-initialized, and move all such logic to `jl_load_dynamic_library_`, which is where the real memory leak was.  I'll prepare a PR to that effect.

**Comment by stevengj** [comment]
I can confirm that the `calloc` workaround fixes the problem for me.

**Comment by rened** [comment]
I just tried https://github.com/garrison/julia/commit/ada46544e3016a247eff895f3ba66be91b9189bb on Linux and still get the same error when I run 
```
Pkg.init(); Pkg.add("HDF5")
```
What does work though is this: 
```
Pkg.init(); Pkg.add("BinDeps"); Pkg.checkout("BinDeps"); Pkg.add("HDF5")
```
So I guess this problem is circumvented in `BinDeps` master?

**Comment by garrison** [comment]
Correct, BinDeps master is also needed.  See the summary [here](https://github.com/JuliaLang/BinDeps.jl/pull/140#issuecomment-93759130).

**Comment by rened** [comment]
Closed via #10832
